server/commands/documentMover.ts

```diff
@@ -38,154 +38,161 @@ async function documentMover({
+    const toIndex =
+      index !== undefined &&
+      document.parentDocumentId === parentDocumentId &&
+      document.collectionId === collectionId &&
+      fromIndex < index
+        ? index - 1
+        : index;
+
+    // Update the properties on the document record
+    document.collectionId = collectionId;
+    document.parentDocumentId = parentDocumentId;
+    document.lastModifiedById = user.id;
+    document.updatedBy = user;
 
-      const newCollection = collectionChanged
-        ? await Collection.scope({
-            method: ["withMembership", user.id],
-          }).findByPk(collectionId, {
-            transaction,
-          })
-        : collection;
+    // Add the document and it's tree to the new collection
+    await newCollection.addDocumentToStructure(document, toIndex, {
+      documentJson,
+      transaction,
+    });
 
-      invariant(newCollection, "collection should exist");
+    if (collection) {
+      result.collections.push(collection);
+    }
 
-      await newCollection?.addDocumentToStructure(document, toIndex, {
-        documentJson,
+    // If the collection has changed then we also need to update the properties
+    // on all of the documents children to reflect the new collectionId
+    if (collectionChanged) {
+      // Reload the collection to get relationship data
+      newCollection = await Collection.scope({
+        method: ["withMembership", user.id],
+      }).findByPk(collectionId, {
+        transaction,
       });
+      invariant(newCollection, "collection should exist");
 
-      if (collection) {
-        result.collections.push(collection);
-      }
+      result.collections.push(newCollection);
 
-      // if collection does not remain the same loop through children and change their
-      // collectionId and move any attachments they may have too. This includes
-      // archived children, otherwise their collection would be wrong once restored.
-      if (collectionChanged) {
-        result.collections.push(newCollection);
-
-        const loopChildren = async (documentId: string) => {
-          const childDocuments = await Document.findAll({
-            where: {
-              parentDocumentId: documentId,
-            },
-          });
-          await Promise.all(
-            childDocuments.map(async (child) => {
-              await loopChildren(child.id);
-              child.collectionId = collectionId;
-              await child.save();
-
-              if (newCollection) {
-                child.collection = newCollection;
-              }
-              result.documents.push(child);
-            })
-          );
-        };
-
-        await loopChildren(document.id);
-
-        const pin = await Pin.findOne({
+      // Efficiently find the ID's of all the documents that are children of
+      // the moved document and update in one query
+      const childDocumentIds = await document.getChildDocumentIds();
+      await Document.update(
```

Comment:
Moved this to a single `UPDATE` instead of a loop of select, update ad infinitum